### PR TITLE
Refactor the tap changer adder interfaces

### PR DESCRIPTION
**Pull request by flomillot**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?**
Refactoring RatioTapChangerAdder and PhaseTapChangerAdder by creating a new common interface TapChangerAdder and doing the same for their StepAdder.

**Does this PR introduce a breaking change or deprecate an API?**
- [x] Yes
- [ ] No

**If yes, please check if the following requirements are fulfilled**

- [x] The *Breaking Change* or *Deprecated* label has been added
- [x] The migration steps are described in the following section

**What changes might users need to make in their application due to this PR? (migration steps)**
If using the interfaces PhaseTapChangerAdder.StepAdder and RatioTapChangerAdder.StepAdder, you need to replace them by PhaseTapChangerStepAdder and RatioTapChangerStepAdder.

### CODE REVIEW

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/PhaseTapChangerStepAdder.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/**
+ * Copyright (c) 2023, All partners of the iTesla project (http://www.itesla-project.eu/consortium)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
```
Comment (olperr1) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/PhaseTapChangerStepAdder.java:6]: This header is deprecated. The one to use is now:
```suggestion
/**
 * Copyright (c) 2024, RTE (http://www.rte-france.com)
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 */
```

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChangerStepAdder.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+/**
+ * Copyright (c) 2023, All partners of the iTesla project (http://www.itesla-project.eu/consortium)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
```
Comment (olperr1) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/RatioTapChangerStepAdder.java:7]: Same here:
```suggestion
/**
 * Copyright (c) 2024, RTE (http://www.rte-france.com)
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 */
```

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TapChangerAdder.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
+/**
+ * Copyright (c) 2023, All partners of the iTesla project (http://www.itesla-project.eu/consortium)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
```
Comment (olperr1) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TapChangerAdder.java:7]:
```suggestion
/**
 * Copyright (c) 2024, RTE (http://www.rte-france.com)
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 */

```

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TapChangerStepAdder.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
+/**
+ * Copyright (c) 2023, All partners of the iTesla project (http://www.itesla-project.eu/consortium)
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
```
Comment (olperr1) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TapChangerStepAdder.java:7]:
```suggestion
/**
 * Copyright (c) 2024, RTE (http://www.rte-france.com)
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 * SPDX-License-Identifier: MPL-2.0
 */

```

**Review by olperr1 — CHANGES_REQUESTED** [review]

File: iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TapChangerStepAdder.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+package com.powsybl.iidm.network;
+
+/**
+ * @author Florent MILLOT {@literal <florent.millot at rte-france.com>}
+ */
+public interface TapChangerStepAdder<S extends com.powsybl.iidm.network.TapChangerStepAdder<S, A>, A extends TapChangerAdder<A, S, ?>> {
```
Comment (olperr1) [inline iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TapChangerStepAdder.java:13]: To be more concise:
```suggestion
public interface TapChangerStepAdder<S extends TapChangerStepAdder<S, A>, A extends TapChangerAdder<A, S, ?>> {
```

Comment (flomillot) [reply iidm/iidm-api/src/main/java/com/powsybl/iidm/network/TapChangerStepAdder.java:14]: I did in #2829 but forgot to do it here, I commit the change

**Review by olperr1 — APPROVED** [review]

**Review by etiennehomer — APPROVED** [review]

**Comment by olperr1** [comment]
@flomillot: Today, we merged a PR (#2758) introducing a regulation mode on `RatioTapChanger` (this explains the conflict currently reported on this PR).
Could you see if it is possible/reasonable to share the methods concerning it in `TapChangerAdder`?

**Comment by flomillot** [comment]
> @flomillot: Today, we merged a PR (#2758) introducing a regulation mode on `RatioTapChanger` (this explains the conflict currently reported on this PR). Could you see if it is possible/reasonable to share the methods concerning it in `TapChangerAdder`?

Yes but actually in the PR it's not even shared in TapChanger, I guess I have to do it also ?

**Comment by olperr1** [comment]
> > @flomillot: Today, we merged a PR (#2758) introducing a regulation mode on `RatioTapChanger` (this explains the conflict currently reported on this PR). Could you see if it is possible/reasonable to share the methods concerning it in `TapChangerAdder`?
> 
> Yes but actually in the PR it's not even shared in TapChanger, I guess I have to do it also ?

@flomillot : Ok, you are right. I propose to keep it like this for now.

**Review by flo-dup — APPROVED** [review]

**Pull request merged**